File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/Python Programs/2_delete_operations.py
class Node:
    def __init__(self, d): self.data, self.next = d, None

class LL:
    def __init__(self): self.head = None
    def ial(self, d): n = Node(d); (setattr(self, 'head', n) if not self.head else [t for t in [self.head] if not [setattr(t, 'next', n) for t in iter(lambda x=self.head: x.next if x.next else None, None)]])
    def df(self): 
        if not self.head: return None
        self.head = self.head.next; return True
    def dl(self):
        if not self.head: return None
        if not self.head.next: self.head = None; return True
        t = self.head; [t := t.next for _ in iter(lambda: t.next.next, None)]; t.next = None; return True
    def da(self, p):
        if not self.head or p < 1: return None
        if p == 1: return self.df()
        t = self.head; [t := t.next for _ in range(p-2) if t and t.next]; 
        if not t or not t.next: return None
        t.next = t.next.next; return True
    def disp(self): 
        if not self.head: print('Empty'); return
        t = self.head; print(' -> '.join([str(t.data)] + [str((t := t.next).data) for _ in iter(lambda: t.next, None)]) + ' -> None')

ll = LL()
[ll.ial(i*10) for i in range(1, 6)]
while True:
    c = input("1:DelFirst 2:DelLast 3:DelAny 4:Show 5:Exit: ")
    if c == '1': ll.df()
    elif c == '2': ll.dl()
    elif c == '3': ll.da(int(input()))
    elif c == '4': ll.disp()
    elif c == '5': break

File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/C Programs/2_delete_operations.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node { int data; struct Node* next; } Node;
Node* cn(int d) { Node* n = malloc(sizeof(Node)); n->data = d; n->next = NULL; return n; }
Node* ial(Node* h, int d) { Node* n = cn(d); if (!h) return n; Node* t = h; while (t->next) t = t->next; t->next = n; return h; }
Node* df(Node* h) { if (!h) return NULL; Node* t = h; h = h->next; free(t); return h; }
Node* dl(Node* h) { if (!h || !h->next) { free(h); return NULL; } Node* t = h; while (t->next->next) t = t->next; free(t->next); t->next = NULL; return h; }
Node* da(Node* h, int p) { if (p == 1) return df(h); Node* t = h; for (int i = 1; i < p-1 && t; i++) t = t->next; if (!t || !t->next) return h; Node* d = t->next; t->next = d->next; free(d); return h; }
void disp(Node* h) { while (h) printf("%d -> ", h->data), h = h->next; printf("NULL\n"); }
int main() {
    Node* h = NULL; int c, p;
    for (int i = 1; i <= 5; i++) h = ial(h, i*10);
    while (scanf("%d", &c) && c != 7) {
        if (c == 1) h = df(h);
        else if (c == 2) h = dl(h);
        else if (c == 3) scanf("%d", &p), h = da(h, p);
        else if (c == 4) disp(h);
    }
    return 0;
}